File: picture-server/src/main/java/com/applory/pictureserver/domain/user/UserController.java
package com.applory.pictureserver.domain.user;

import com.applory.pictureserver.domain.favorite.Favorite;
import com.applory.pictureserver.domain.favorite.FavoriteDTO;
import com.applory.pictureserver.domain.favorite.FavoriteService;
import com.applory.pictureserver.domain.user.querydto.SellerListVM;
import com.applory.pictureserver.shared.Result;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.UUID;

@RestController
@RequestMapping("/api/v1/users")
@RequiredArgsConstructor
public class UserController {

    private final UserService userService;

    private final FavoriteService favoriteService;

    @PostMapping("")
    @ResponseStatus(HttpStatus.CREATED)
    public UserDto.VM createUser(@Valid @RequestBody UserDto.Create dto) {
        User newUser = userService.createUser(dto);
        return new UserDto.VM(newUser);
    }

    @GetMapping("/check-nickname")
    public void checkNickname(@RequestParam String nickname) {
        userService.checkNickname(nickname);
    }

    @GetMapping("/me")
    public Result getUserMe() {
        return Result.success(userService.getUserMe());
    }

    @GetMapping("/seller")
    public Result getSellerUsers(@Valid UserDto.SearchSeller search, Pageable pageable) {
        return Result.success(userService.getSellerUsers(search, pageable));
    }

    @GetMapping("/client")
    public Page<UserDto.VM> getClientUsers(@Valid UserDto.SearchClient search, Pageable pageable) {
        return userService.getClientUsers(search, pageable).map(UserDto.VM::new);
    }

    @GetMapping("/seller/{id}")
    public UserDto.SellerVM getSellerUser(@PathVariable UUID id, @RequestParam(required = false) UUID requesterId) {
        return userService.getSellerDetail(id, requesterId);
    }

    @PostMapping("/{id}/profile-image")
    public Result updateProfileImage(@PathVariable("id") UUID userId, UserDto.UpdateProfileImage dto) {
        userService.updateProfileImage(userId, dto);
        return Result.success();
    }

    @PostMapping("/favorites")
    public Result toggleFavorite(@RequestBody FavoriteDTO.Toggle body) {
        favoriteService.toggleFavorite(body);
        return Result.success();
    }
}

File: picture-server-kt/src/main/kotlin/com/applory/pictureserverkt/user/UserController.kt
package com.applory.pictureserverkt.user

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import javax.validation.Valid

@RestController
@RequestMapping("/api/v1/users")
class UserController(private val userService: UserService) {

    @PostMapping("")
    @ResponseStatus(HttpStatus.CREATED)
    fun createUser(@Valid @RequestBody dto: UserDto.Create): UserDto.VM {
        return UserDto.convertToVM(userService.createUser(dto))
    }

    @GetMapping("/me")
    fun getUserMe(): UserDto.VM {
        return UserDto.convertToVM(userService.getUserMe())
    }

    @GetMapping("/client")
    fun getClientUsers(search: UserDto.SearchClient, pageable: Pageable): Page<UserDto.VM> {
        return userService.getClientUsers(search, pageable).map { UserDto.convertToVM(it) }
    }

    @GetMapping("/seller")
    fun getSellerUsers(@Valid search: UserDto.SearchSeller, pageable: Pageable): Page<UserDto.VM> {
        return userService.getSellerUsers(search, pageable).map { UserDto.convertToVM(it) }
    }

}
